Add first draft of ZenIP 42202 for non-ceasable sidechains (#17)
HorizenOfficial/ZenIPs: zenip-42202.md

# ZenIP-42202

    ZenIP: 42202
    Title: Non-ceasable Sidechains
    Owner: Paolo Tagliaferri, <paolotagliaferri@horizenlabs.io>; cronic, <cronic@horizenlabs.io>
    Discussions-To: Paolo Tagliaferri, <paolotagliaferri@horizenlabs.io>
    Comments-URI: https://horizen.global/invite/discord
    Status: Draft
    Type: Consensus
    Created: 2022-08-01
    License: MIT

## Table of Contents

<!--ts-->

- [Abstract](#abstract)
- [Motivation](#motivation)
- [Specification](#specification)
  - [Constraints](#constraints)
  - [Maturity](#maturity)

## Abstract

> This document describes the introduction of a new type of sidechains that “never cease” from processing transactions. The version for this type of sidechains is “v2”.

Reviewer: "This document introduces a new type of sidechain that never ceases from processing transactions."

Author: Thank you @zain-cheng, I've applied your suggestion!

> Currently, all sidechains must specify a `withdrawalEpochLength` at creation time, and such value immediately determines when a submission window opens and closes for each epoch. 

Reviewer 2: ```suggestion
With sidechains v0 and v1, all sidechains must specify a `withdrawalEpochLength` > 0 at time of creation, this value determines the length of the epoch and when a submission window opens and closes for each epoch. 
```

At the beginning of any epoch, except for the first one, there is a subset of blocks that constitutes the so-called **submission window**, in which the sidechain is required to publish at least one certificate for the previous epoch. Any certificate received outside this window is automatically rejected.

Certificates act as an heartbeat of a sidechain and are the only way to move coins back to the mainchain.

If a sidechain doesn’t provide at least one certificate for an epoch, it is considered as “ceased” and no more transactions to and from the sidechain are accepted (only *Ceased Sidechain Withdrawals* are allowed as an emergency procedure to recover funds).

![A ceasable sidechain](ZenIP-42202/ZenIP-42202-1.png)

We propose to introduce sidechains v2 that are not required to submit certificates during a submission window but can continue to process transactions without the risk of ceasing.

Please note that this non-ceasing behavior is not enabled for all sidechains v2, but it must be explicitly requested by setting the `withdrawalEpochLength` to “0”.

> To recap, sidechains v2 will have two different behaviors based on the creation parameters:

Reviewer 2: ```suggestion
Sidechains v2 will have two different behaviors based on the creation parameters:
```

> - if the `withdrawalEpochLength` is set to “0”, the sidechain never cease

Reviewer 2: ```suggestion
- if the `withdrawalEpochLength` is set to “0”, the sidechain can never cease
```

Reviewer 3: Using the word, "can" as a modal verb in this sentence gives a fact about the sidechain. Not using "can" makes the sentence states a definite action or reaction about the sidechain....it "never ceases" period.

Author: Thanks @Ramone-HL, I've just removed `can`, the new sentence is:

`if the withdrawalEpochLength is set to “0”, the sidechain never ceases`.

- if the `withdrawalEpochLength` is not “0”, the sidechain works exactly as sidechains v1 (with normal ceasing behavior)

## Motivation

> The entire sidechain mechanism that has been included in the Zen blockchain under the name of “Zendoo“ was designed and implemented having in mind the security of users in a fully decentralized scenario.

Reviewer 2: ```suggestion
The sidechain mechanism that has been included in the Horizen blockchain under the name of “Zendoo“ was designed and implemented having in mind the security of users in a fully decentralized scenario.
```

> In the meanwhile, some sidechains have started to be implemented in semi-centralized environments, where security constraints are way more relaxed.

Reviewer 2: ```suggestion
In the meanwhile, use cases with more relaxed security constraints that can benefit from a semi-centralized environment have been found.
```

Reviewer 3: There are some use cases where sidechains are implemented in a semi-centralized environment with relaxed security constraints.

Author: Done.

> For instance, in a fully decentralized sidechain, the `Mainchain Backward Transfer Request` (MBTR) guarantees to users the possibility to perform withdrawal of funds directly from mainchain in case the sidechain is somehow compromised (but continues to publish certificates). Instead, if the sidechain ceases, users can still retrieve their funds by using the `Ceased Sidechain Withrawal` (CSW) mechanism.

Reviewer 2: ```suggestion
In a fully decentralized sidechain, the `Mainchain Backward Transfer Request` (MBTR) guarantees to users the possibility to perform withdrawal of funds directly from mainchain in case the sidechain is somehow compromised (but continues to publish certificates). If the sidechain ceases, users can still retrieve their funds by using the `Ceased Sidechain Withrawal` (CSW) mechanism.
```

- **Sidechain Creation**

  - `withdrawalEpochLength` must be set to 0 (to disable the fixed-length epochs and submission windows, otherwise they work exactly as sidechains v1 do).

  - `mainchainBackwardTransferRequestDataLength` must be set to 0 (to disable the MBTR).

  - `wCeasedVk` must be empty (to disable the CSW).

- **Certificates**

  - The `quality` field must always be set to “0”.

  - If mainchain receives first a certificate of epoch N (either in mempool or in a block) and then another certificate of same epoch, the latter is rejected.

  - All certificates must refer (through the `endEpochCumScTxCommTreeRoot` field) to different ordered block. So, for instance, if the mainchain receives a certificate for epoch N referring to the mainchain block M (i.e. setting the `endEpochCumScTxCommTreeRoot `field to the same value specified in the header of the mainchain block M), any certificate coming next must refer to a `endEpochCumScTxCommTreeRoot` (and thus to a block) with higher height than M (M + 1, M + 2, etc., are all good candidates). Such constraint enforces that a certificate of higher epoch refers to a more updated state of the mainchain and of the sidechain.

>   - More than one certificate for the same sidechain can be included in the same block (if they respect all the constraints above).

Reviewer 2: ```suggestion
  - More than one certificate for the same sidechain can be included in the same block (if they respect all the constraints above and are ordered sequentially).
```

When sidechains v2 are created with `withdrawalEpochLength` not set to 0, they will enable the default ceasing mechanism as it is intended for sidechains v1.

### Maturity

For sidechain v0 and v1, Zendoo implements a maturity mechanism for the amount of coins transferred to and from any sidechain.

> For instance, whenever transactions are created with a cross-chain output (sidechain creation, forward transfer, or mainchain backward transfer request), the coins and fees are able to move to the sidechain but the amount is not immediately added to its balance.  The coins are marked as immature and after some block confirmation, they are included in the sidechain balance. The number of confirmations is set to `10` blocks for mainnet in the chain parameters.

Reviewer 2: @ptagl Please add a sentence on SC value pool (only balance that moved into SC can be moved out of it). Also does immaturity of FTs mean that if you forward transfer before the epoch end, you cannot backward transfer the same coins in the submission window as long as the FT has < 10 confirmations?

> The coins are marked as immature and after some block confirmation,

A little ambiguous, better `Cross-chain outputs are marked as immature on the mainchain until they have been confirmed by some mainchain blocks, and only after maturing they are added to the sidechain balance.`

> The number of confirmations is set to `10` blocks for mainnet in the chain parameters.

`Cross-chain outputs mature after 10 confirmations on mainnet, this is set in the chain parameters`

Author: Added a short description about how the balance is used by mainchain.

> The same concept applies to backward transfers in a certificate. During the submission window phase, a certificate can be superseded by another certificate of higher quality. In this case, it makes sense to mark the backward transfers as immature until the certificate is final.

Reviewer 2: ```suggestion
The concept of maturity also applies to backward transfers in a certificate. During the submission window phase, a certificate can be superseded by another certificate of higher quality. Because of this backward transfers included in a certificate are immature until the certificate is final.
```

> Moreover, a certificate can be reverted back, just in case a sidechain ceases. Zendoo considers maturity of any backward transfer included in a top quality certificate for epoch N only after the end of the submission window for epoch N + 1.

Reviewer 2: ```suggestion
In case a sidechain ceases any backward transfers in an immature certificate are reverted. Zendoo only considers backward transfers included in a top quality certificate for epoch N mature at the end of the submission window for epoch N + 1 (another certificate was submitted in submission window of epoch N + 1).
```

Reviewer 3: If a sidechain ceases, any backward transfers in an immature certificate are reverted. Zendoo only considers backward transfers included in a top quality certificate for epoch N mature at the end of the submission window for epoch N + 1 (submitted another certificate in submission window of epoch N + 1).

Author: Thanks @Ramone-HL, I've applied your suggestion, I only expanded a bit the text between brackets:

`(once a new certificate is submitted during the submission window of epoch N + 1)`.

> Since non-ceasable sidechains don’t support the CSW, such maturity concept is not useful. Therefore, the maturity is going to be ignored and every cross-chain output or backward transfer will be immediately considered spendable (and added/removed to the sidechain balance). This only applies to sidechains v2 with `withdrawalEpochLength` set to “0”.

Reviewer 2: ```suggestion
Since non-ceasable sidechains don’t support CSW, the concept of maturity is unnecessary. Therefore, every cross-chain output or backward transfer is immediately considered spendable (and added/removed to the sidechain balance). This only applies to sidechains v2 with `withdrawalEpochLength` set to “0”.
```

Reviewer 3: Since non-ceasable sidechains don’t support CSW, the concept of maturity is ignored.

# ZenIP-42202

    ZenIP: 42202
    Title: Non-ceasable Sidechains
    Owner: Paolo Tagliaferri, <paolotagliaferri@horizenlabs.io>; cronic, <cronic@horizenlabs.io>
    Discussions-To: Paolo Tagliaferri, <paolotagliaferri@horizenlabs.io>
    Comments-URI: https://horizen.global/invite/discord
    Status: Draft
    Type: Consensus
    Created: 2022-08-01
    License: MIT

## Table of Contents

<!--ts-->

- [Abstract](#abstract)
- [Motivation](#motivation)
- [Specification](#specification)
  - [Constraints](#constraints)
  - [Maturity](#maturity)

## Abstract

This document introduces a new type of sidechain that never ceases from processing transactions. The version for this type of sidechains is “v2”.

> With sidechains v0 and v1, all sidechains must specify a `withdrawalEpochLength` > 0 at time of creation, this value determines the length of the epoch and when a submission window opens and closes for each epoch. 

Reviewer 4: ```suggestion
For sidechains v0 and v1, a `withdrawalEpochLength` > 0 must be specified at the time of creation. This value determines the length of the epoch and when a submission window opens and closes for each epoch. 
```

Reviewer 5: I use this application and plugin that is quite good at correcting grammar mistakes: https://grammarlysoftware.com/download you may want to take a look at it 😃

Author: @otoumas I always use it too, in addition for this ZenIP (and the ones coming in the future) I guess we can also rely on @Ramone-HL for final review and checks.

> The life cycle of any sidechain is composed of epochs, that are sets of consecutive blocks whose size is determined by the `withdrawalEpochLength`. The first epoch begins with the block in which the sidechain is created (through a transaction containing its SidechainCreationOutput), the second one starts after `withdrawalEpochLength` blocks and so on.

Reviewer 4: ```suggestion
The life cycle of any sidechain is composed of epochs, which are sets of consecutive blocks whose size is defined by the `withdrawalEpochLength`. The first epoch begins with the block that creates the sidechain, the second epoch starts after a `withdrawalEpochLength` number of blocks, and so on.
```

Reviewer 4: I don't think that it is necessary to include `SidechainCreationOutput` in this sentence.

> At the beginning of any epoch, except for the first one, there is a subset of blocks that constitutes the so-called **submission window**, in which the sidechain is required to publish at least one certificate for the previous epoch. Any certificate received outside this window is automatically rejected.

Reviewer 4: ```suggestion
At the beginning of every epoch (except for the first one), there is a subset of blocks that constitutes the **submission window**, during which the sidechain is required to publish at least one certificate for the previous epoch. Any certificate received outside this window is automatically rejected.
```

> Certificates act as an heartbeat of a sidechain and are the only way to move coins back to the mainchain.

Reviewer 4: ```suggestion
Certificates are the only way to move coins back to the mainchain.
```

> If a sidechain doesn’t provide at least one certificate for an epoch, it is considered as “ceased” and no more transactions to and from the sidechain are accepted (only *Ceased Sidechain Withdrawals* are allowed as an emergency procedure to recover funds).

Reviewer 4: ```suggestion
If a sidechain doesn’t provide at least one certificate for an epoch, it is considered “ceased” and no more transactions to or from the sidechain are accepted. Only *Ceased Sidechain Withdrawals* are allowed as an emergency procedure to recover funds.
```

![A ceasable sidechain](ZenIP-42202/ZenIP-42202-1.png)

> We propose to introduce sidechains v2 that are not required to submit certificates during a submission window but can continue to process transactions without the risk of ceasing.

Reviewer 4: ```suggestion
The proposal is to introduce sidechains v2 that are not required to submit certificates during a submission window and can continue to process transactions without the risk of ceasing.
```

> Please note that this non-ceasing behavior is not enabled for all sidechains v2, but it must be explicitly requested by setting the `withdrawalEpochLength` to “0”.

Reviewer 4: ```suggestion
This non-ceasing behavior is not enabled for all sidechains v2. It must be explicitly defined by setting the `withdrawalEpochLength` to “0”.
```

Sidechains v2 will have two different behaviors based on the creation parameters:

- if the `withdrawalEpochLength` is set to “0”, the sidechain never ceases

> - if the `withdrawalEpochLength` is not “0”, the sidechain works exactly as sidechains v1 (with normal ceasing behavior)

Reviewer 4: ```suggestion
- if the `withdrawalEpochLength` is greater than “0”, the sidechain works exactly as sidechains v0 and v1 with ceasing behavior
```

Reviewer 4: Adding `v0` here because it is also mentioned above.

> - if the `withdrawalEpochLength` is not “0”, the sidechain works exactly as sidechains v1 (with normal ceasing behavior)

Reviewer 5: This looks like a crucial decision when a sidechain is declared, what do you think if a specific configuration parameter is set instead of reusing the `withdrawalEpochLength`? Maybe a specific boolean environment variable: `ceasable` could be set? This way it explicitly defines the sidechain as ceasable/non-ceasable , instead of implicitly by the value of `withdrawalEpochLength`

Author: We took into account the possibility to add a new boolean in the `SidechainCreationOutput` but in the end we preferred to avoid for the following reasons:
- the new boolean would be strictly connected to the value of `withdrawalEpochLength`, meaning that if the sidechain is `ceasable` than the `withdrawalEpochLength` must be greater than zero, otherwise, for non-ceasable sidechains, only zero is accepted as valid. In this sense, adding a new field doesn't provide more flexibility but at the same time would bring more complexity as the validation of a request should take into account two fields instead of one
- adding a new boolean would add extra space to the serialization of the `SidechainCreationOutput`; even though it would be minimal, we always try to minimize the footprint of the blockchain
- when creating a sidechain, thinking about the value to be set for the `withdrawalEpochLength` is always crucial, also for ceasable sidechains, as it determines the frequency of certificates to be published. Moreover, when creating a non-ceasable sidechain, the validation rules always enforce that CSW and MBTR are disabled, so it seems unlucky to create a sidechain with wrong behavior by mistake.
- with a new boolean, we should also change the interface of the RPC command to create sidechains, making unusable any existing script or tools used by other teams unless they get properly updated

Reviewer 5: I see, thanks for the explanation! 
What about making `withdrawalEpochLength` optional? so that it is only added to the `SidechainCreationOutput` if it is a ceasable sidechain?

Author: You're welcome!

Currently the RPC command `sc_create` considers `withdrawalEpochLength` as an optional field, whose default value is 100 (that is the minimum epoch length for ceasable sidechains).
Changing such behavior could break the compatibility with tools and scripts that are eventually calling `sc_create` without explicitly setting the `withdrawalEpochLength`.
Making the change only to the `SidechainCreationOutput` could cause similar issues, as tools like the Explorer and the Bootstrapping Tool might be implemented to always parse the `withdrawalEpochLength` from the creation transaction. In general, we try to reduce as much as possible changes on the mainchain that would affect other software.
There is also a serialization issue here: if we make that field optional, we wouldn't have a way to understand whether the `withdrawalEpochLength` has to be read or not. To make things even more complex, in the past, when we introduced sidechains v1, we had to use the first byte of the `withdrawalEpochLength` to specify the sidechain version, so making changes to the serialization of this field would have impacts also on the `version` field.

Reviewer 5: I see! Ok sounds like allowing 0 value is the best option then 😄

## Motivation

> The sidechain mechanism that has been included in the Horizen blockchain under the name of “Zendoo“ was designed and implemented having in mind the security of users in a fully decentralized scenario.

Reviewer 4: ```suggestion
The sidechain mechanism introduced in the Horizen blockchain with “Zendoo“ was designed and implemented for a fully decentralized scenario.
```

> There are some use cases where sidechains are implemented in a semi-centralized environment with relaxed security constraints.

Reviewer 4: ```suggestion
However, there are some use cases where sidechains are implemented in a semi-centralized environment with relaxed security constraints.
```

> In a fully decentralized sidechain, the `Mainchain Backward Transfer Request` (MBTR) guarantees to users the possibility to perform withdrawal of funds directly from mainchain in case the sidechain is somehow compromised (but continues to publish certificates). If the sidechain ceases, users can still retrieve their funds by using the `Ceased Sidechain Withrawal` (CSW) mechanism.

Reviewer 4: ```suggestion
In a fully decentralized sidechain, the `Mainchain Backward Transfer Request` (MBTR) guarantees the withdrawal of funds directly from mainchain in case the sidechain is compromised (but continues to publish certificates). Additionally, if the sidechain ceases, funds can be withdrawn with the `Ceased Sidechain Withrawal` (CSW). In a semi-centralized sidechain, however, MBTR and CSW are not necessary.
```

> In semi-centralized environments, where users can trust the actors running the sidechain, those withdrawal mechanisms (MBTR and CSW) are not needed and users can rely on the normal Backward Transfer mechanism triggered on the sidechain itself. Moreover, in these environments, having fixed certificate submission windows may lead to unwanted or unexpected ceasing due to a variety of reasons:

Reviewer 4: ```suggestion
In semi-centralized environments, the actors running the sidechain can be trusted, and users can rely on the Backward Transfer mechanism triggered on the sidechain itself. Moreover, having fixed certificate submission windows may lead to unwanted or unexpected ceasing due to a variety of reasons:
```

- bugs in the software running the nodes

- connectivity issues

- congestion of blocks (too many transactions and certificates for one single block)

> **Note**: The changes proposed in this ZenIP can mitigate such risks without affecting the security of the sidechain.

Reviewer 4: ```suggestion
The proposed changes can mitigate such risks without affecting the security of the sidechain.
```

## Specification

The creation of sidechains v2 is triggered by the consensus rules starting with the activation of the fork point #10.

Any attempt to create a sidechain v2 on mainnet before that height will be rejected.

> Sidechains v2 will keep all the features included in v1 in addition to the possibility to enable the non-ceasing behavior by properly setting the `withdrawalEpochLength`.

Reviewer 4: ```suggestion
Sidechains v2 includes all the features in sidechains v1, and adds the possibility of enabling non-ceasing behavior by setting `withdrawalEpochLength = "0"`.
```

![Non-ceasable sidechain](ZenIP-42202/ZenIP-42202-1.png)

### Constraints

> Sidechains v2 will never cease when setting `withdrawalEpochLength` to “0” at creation time. They are not forced to publish certificates during fixed submission windows. However, they can publish a certificate whenever they want, with the only two following constraints:

Reviewer 4: ```suggestion
Sidechains v2 will never cease when setting `withdrawalEpochLength` to “0” at creation time. They are not forced to publish certificates during fixed submission windows. They can publish a certificate whenever they want, with the two following constraints:
```

- The `epoch` values of certificates must be consecutive (i.e. after a certificate of epoch N the sidechain must publish a certificate of epoch N + 1)

> - The `quality` mechanism is disabled, so it’s not possible to publish two or more certificates for the same epoch

Reviewer 5: What is the reason for this constraint? (just curious)

Author: If you're referring to the continuity of epochs, it's been done to simplify the implementation for both mainchain and sidechain: on mainchain side, it's easier, for instance, to revert a certificate of epoch E (in case of disconnection of a block) by restoring the certificate of epoch E-1; on sidechain side, allowing to completely skip epochs would require merging the proof for an epoch E with the proof of epoch E+1 and it is currently not possible

If you're referring to the `quality` field, it's been set to 0 to make more evident that the sidechain is working with a different behavior (ceasable sidechains can only provide certificates with quality greater than zero by SDK design). Regardless of the value, it's good and also necessary to disable the "quality" mechanism so that certificates are immediately "final" upon inclusion into a block and the mainchain shouldn't deal with superseding and maturity.

Reviewer 5: Yes sorry I was referring to the quality one 😅 , how is the quality disabled? depending on the value of `withdrawalEpochLength` ?

Author: Correct, using `version: 2` and `withdrawalEpochLength: 0` when creating a sidechain, we know that it is non-ceasable and in such case we only accept certificates with quality set to 0.

These non-ceasable sidechains must satisfy the following constraints:

- **Sidechain Creation**

>   - `withdrawalEpochLength` must be set to 0 (to disable the fixed-length epochs and submission windows, otherwise they work exactly as sidechains v1 do).

Reviewer 4: Keep consistent, either `"0"` or `0`

Author: Fixed.

>   - `withdrawalEpochLength` must be set to 0 (to disable the fixed-length epochs and submission windows, otherwise they work exactly as sidechains v1 do).
> 
>   - `mainchainBackwardTransferRequestDataLength` must be set to 0 (to disable the MBTR).
> 
>   - `wCeasedVk` must be empty (to disable the CSW).

Reviewer 4: ```suggestion
  - `withdrawalEpochLength` must be set to 0. 
  This disabled fixed-length epochs and submission windows.

  - `mainchainBackwardTransferRequestDataLength` must be set to 0. 
  This disables MBTR.

  - `wCeasedVk` must be empty. 
  This disables CSW.
```

- **Certificates**

  - The `quality` field must always be set to “0”.

>   - If mainchain receives first a certificate of epoch N (either in mempool or in a block) and then another certificate of same epoch, the latter is rejected.

Reviewer 4: ```suggestion
  - If the mainchain first receives a certificate of epoch N (either in the mempool or in a block) and then receives another certificate in same epoch N, the latter is rejected.
```

  - All certificates must refer (through the `endEpochCumScTxCommTreeRoot` field) to different ordered block. So, for instance, if the mainchain receives a certificate for epoch N referring to the mainchain block M (i.e. setting the `endEpochCumScTxCommTreeRoot `field to the same value specified in the header of the mainchain block M), any certificate coming next must refer to a `endEpochCumScTxCommTreeRoot` (and thus to a block) with higher height than M (M + 1, M + 2, etc., are all good candidates). Such constraint enforces that a certificate of higher epoch refers to a more updated state of the mainchain and of the sidechain.

>   - More than one certificate for the same sidechain can be included in the same block (if they respect all the constraints above and are ordered sequentially).

Reviewer 4: ```suggestion
  - More than one certificate for the same sidechain can be included in the same block if they follow all the constraints above and are ordered sequentially.
```

> When sidechains v2 are created with `withdrawalEpochLength` not set to 0, they will enable the default ceasing mechanism as it is intended for sidechains v1.

Reviewer 4: Do we need to say this here again? It was already mentioned above.

> ### Maturity

Reviewer 4: Because this is a proposal for non-ceasing sidechains, instead of describing the concept of maturity for ceasing sidechains, we could include a link to a page that defines what it means.

Then for the "Maturity" section, I would only keep the last part that is related to non-ceasing sidechains.

Author: Unfortunately, we don't have at the moment a document ready to describe maturity and we haven't even agreed about the place to be used for mainchain documentation. I guess that until the decision is made, the only possibility is to briefly describe the maturity concept in the ZenIP itself.

Reviewer 4: That is fine.
I cannot resolve my comments. Feel free to resolve any unclosed ones.

For sidechain v0 and v1, Zendoo implements a maturity mechanism for the amount of coins transferred to and from any sidechain.

> For instance, whenever transactions are created with a cross-chain output (sidechain creation, forward transfer, or mainchain backward transfer request), the coins and fees are able to move to the sidechain but the amount is not immediately added to its balance. The coins are marked as immature and after some block confirmation, they are included in the sidechain balance. The number of confirmations is set to `10` blocks for mainnet in the chain parameters.
> How is this concept of maturity handled by the mainchain? Whenever a transaction including a CSW or a certificate with BTs is received, the validation rules enforce that the total amount of coins moved back from sidechain to mainchain does not exceed the sidechain balance. If a sidechain tries to move back more coins than its balance, the transaction is rejected.

Reviewer 4: ```suggestion
This means that whenever transactions are created with a cross-chain output (sidechain creation, forward transfer, or mainchain backward transfer request), the coins and fees are transferred to the sidechain but the amount is not immediately added to its balance. The coins are marked as immature, and only after some block confirmations, are they included in the sidechain balance. The number of confirmations is set to `10` blocks for mainnet in the chain parameters.
On the mainchain, whenever a transaction including a CSW or a certificate with BTs is received, the validation rules enforce that the total amount of coins transfer from sidechain to mainchain does not exceed the sidechain balance. If a sidechain tries to transfer more coins than its balance, the transaction is rejected.
```

> The concept of maturity also applies to backward transfers in a certificate. During the submission window phase, a certificate can be superseded by another certificate of higher quality. Because of this backward transfers included in a certificate are immature until the certificate is final.

Reviewer 4: ```suggestion
The concept of maturity also applies to backward transfers in a certificate. During the submission window phase, a certificate can be superseded by another certificate of higher quality. For this reason, backward transfers included in a certificate are immature until the certificate is finalized.
```

> 

Reviewer 5: Why is this mechanism needed for ceasable sidechains, and not for non-ceasable ones?

Author: This is due to the quality mechanism: since ceasable sidechains can publish several certificates with increasing quality for a given epoch, we need to wait until the last one is published until we agree on the final state of the sidechain. Moreover, in case of ceasing, the last certificate is "discarded" and the sidechain is restored to previous state, so, to consider as "final" a certificate of epoch E, we need to wait until the end of the submission window for epoch E+1.

Consider also that for non-ceasable sidechains there is no fixed epoch length and there is no submission window.

Reviewer 5: Ok, so if I understand this correctly, for Non-ceasable sidechains, as there is no limit (epoch or submission window) to receive a certificate from a sidechain, the only requirement is that the mainchain block referenced in that certificate is higher than the previous certificate block, so sending a second certificate for the same epoch and higher quality would break this rule. 
But then what is the benefit the quality mechanism provides for ceasable sidechains? And why is this benefit not provided for non-ceasable ones?

Author: As far as I know, there are at least two advantages when talking in general about the [Latus design](https://arxiv.org/pdf/2103.13754.pdf):
- in case of sidechain forks happening during the submission window, a part of the network might agree on a state S1 and submit a certificate C1 while another part might agree on S2 and send the certificate C2. The mainchain has no way to detect which state (and certificate) is legit, so keeping the quality of certificates proportional to the length of the chain it refers to, makes the mainchain been able to choose the best certificate (higher quality => certificate created from the state of the longest chain)
- if a malicious actor partially takes control of a sidechain, it can forge a sidechain block to get the longest chain (without publishing it to other peers) and then submit a certificate with highest quality. In this way, other honest peers wouldn't be able to continue forging new blocks as they don't know the block forged by the attacker and the state is already commited to mainchain. To prevent such attacks, quality is proportional to the number of nodes that signed the certificate: having more signers means that the quality is higher and that the certificate has higher chances to be considered as final by the mainchain (reducing the possibility for attackers to break the sidechain).

I'm not sure if both behaviors are currently taken into account in Blaze SDK and, if this is the case, if there is any difference between the design described in the paper for Latus and the current implementation.

Reviewer 5: I see, and the reason why we consider this not to be a problem in non-ceasable sidechains is that we consider these ones to be used only in private/centralized environments mainly, right?

Author: Yes, at least for the malicious scenario.

For the possibility of forks, we don't have such issue anymore since we disabled the submission window, so the sidechain can take its time to generate the certificate and publish it only when it's sure that the state being committed is final.

If a sidechain ceases, any backward transfers in an immature certificate are reverted. Zendoo only considers backward transfers included in a top quality certificate for epoch N mature at the end of the submission window for epoch N + 1 (once a new certificate is submitted during the submission window of epoch N + 1).

> Since non-ceasable sidechains don’t support CSW, the concept of maturity is unnecessary. Therefore, every cross-chain output or backward transfer is immediately considered spendable (and added/removed to the sidechain balance). This only applies to sidechains v2 with `withdrawalEpochLength` set to “0”.

Reviewer 4: ```suggestion
Since non-ceasable sidechains don’t support CSW, the concept of maturity is unnecessary. This means that every cross-chain output or backward transfer is immediately considered spendable and it is immediately added to or removed from the sidechain balance.
```
